src/rust/engine/remote_provider/remote_provider_reapi/src/byte_store.rs

```diff
@@ -390,6 +444,139 @@ impl ByteStoreProvider for Provider {
+            current_size += message_size;
+        }
+
+        if !current_chunk.is_empty() {
+            chunks.push(current_chunk);
+        }
+
+        let client = self.cas_client.as_ref();
+        let destination = Arc::new(Mutex::new(destination));
+
+        let chunk_futures = chunks
+            .clone()
+            .into_iter()
+            .map(|chunk| {
+                let digests = chunk
+                    .iter()
+                    .map(|d| d.into())
+                    .collect::<Vec<remexec::Digest>>();
+                let request = remexec::BatchReadBlobsRequest {
+                    instance_name: self.instance_name.as_ref().cloned().unwrap_or_default(),
+                    digests: digests,
+                    acceptable_compressors: vec![],
+                };
+
+                let client = client.clone();
+                let destination = destination.clone();
+
+                retry_call(
+                    client,
+                    move |mut client, _| {
+                        let request = request.clone();
+                        let chunk = chunk.clone();
+                        let destination = destination.clone();
+
+                        async move {
+                            let response = client.batch_read_blobs(request).await?;
+                            let responses =
+                                Provider::validate_batch_response(chunk, response.into_inner())
+                                    .map_err(Status::unknown)?;
+
+                            if let Some(err) = responses.values().find_map(|res| res.as_ref().err())
+                            {
+                                return Err(err.clone());
+                            }
+
+                            let to_write = responses
+                                .into_iter()
+                                .map(|(d, res)| (d, res.unwrap()))
+                                .collect::<Vec<_>>();
+
+                            let digests = to_write.iter().map(|(d, _)| (*d)).collect::<Vec<_>>();
+
+                            destination
+                                .lock()
+                                .await
+                                .write(to_write)
+                                .await
+                                .map_err(Status::unknown)?;
+
+                            Ok(digests)
+                        }
+                    },
+                    status_is_retryable,
+                )
+            })
+            .collect::<Vec<_>>();
+
+        let chunk_results = futures::future::join_all(chunk_futures).await;
+
+        let mut result = HashMap::new();
+
+        for (response, digests) in chunk_results.into_iter().zip(chunks) {
+            match response {
+                Ok(digests) => {
+                    for digest in digests {
+                        result.insert(digest, Ok(true));
+                    }
+                }
+                Err(err) => {
```

Comment:
On failure we don't know which digests failed in the batch right now, so they all get the same status, which seems less than ideal. Open to suggestions how I can improve this